server/server.py

```diff
@@ -0,0 +1,566 @@
+    - Generates an AES key, encrypts it with client's public RSA key, and sends the encrypted AES key back.
+    - Updates the database with the new AES key.
+    """
+    aes_key = aes.generate_key()  # Generate a new AES key for the client
+    clients_dict[client_name].aes_key = aes_key
+
+    try:
+        # Encrypt AES key using client's RSA public key
+        enc_aes_key = rsa.encrypt(aes_key, clients_dict[client_name].rsa_public_key)
+    except:
+        if not is_reg:
+            # Issue with encryption, send reconnection failure if it's not during registration
+            print("Had an error while encrypting AES key with RSA public key. Possibly bad RSA key.")
+            failed_recon_pckt = ResourceWarning.ReconnectionDeclinedPacker(SERVER_VER, clients_dict[client_name].id)
+            conn.sendall(failed_recon_pckt)
+            return ERROR
+        return GENERAL_ERROR
+
+    # Select the appropriate packet type based on registration or reconnection
+    if is_reg:
+        pubkey_pckt = ResponsePacker.EncryptedAESKeyRegistrationPacker(SERVER_VER, clients_dict[client_name].id, enc_aes_key)
+    else:
+        pubkey_pckt = ResponsePacker.EncryptedAESKeyReconnectionPacker(SERVER_VER, clients_dict[client_name].id, enc_aes_key)
+    send_packet(conn, pubkey_pckt.pack())
+
+    # Update the database with the newly generated AES key
+    with sqlite3.connect("defensive.db") as conn_db:
+        cursor = conn_db.cursor()
+        cursor.execute("UPDATE clients SET AES_key = ? WHERE ID = ?", (aes_key, clients_dict[client_name].id))
+        conn_db.commit()
+
+    print(f"Public key accepted, encrypted AES key sent.")
+    return SUCCESS
+
+# **********************************************************************
+# File recv related functions
+# **********************************************************************
+def get_enc_file_content(conn, file_name, total_packets, enc_file_size, file_size, enc_file_content):
+    """
+    Receives the encrypted file content by collecting packets from the connection.
+    Validates each packet's code, packet number, and file name for integrity.
+    
+    Args:
+        conn: The connection object.
+        file_name (str): The expected name of the file being received.
+        total_packets (int): Total number of packets expected.
+        enc_file_size (int): Expected size of the encrypted file.
+        file_size (int): Expected size of the decrypted file.
+        enc_file_content (str): Placeholder for accumulating file content.
+        
+    Returns:
+        str: The full encrypted file content if successful; otherwise, 0 on error.
+    """
+    for curr_packet in range(2, total_packets + 1): 
+        # Receive the next packet and validate essential fields
+        packet = recv_packet(conn)
+        if packet['code'] != 828:
+            print(f"Error: Invalid packet code: ({packet['code']} instead of 828)")
+            return 0
+        if packet['packet_number'] != curr_packet:

```

Comment:
extreme code duplication - make some function to get and assert a certain packet